### ISSUE TITLE
Покупатель редактирует корзину

### ISSUE THREAD

**Issue by markshevchenko** [progmsk/store#6]
Как *покупатель* я могу изменить количество книг в корзине или удалить их, чтобы я мог исправить случайную ошибку или передумать.

## Основной сценарий
1. Покупатель видит список заказанных книг, их количество и общую сумму.
2. Покупатель изменяет количество экземпляров книги в корзине.
3. Покупатель удаляет книгу из корзины.
4. Сайт пересчитывает сумму заказа.

**Comment by lukas-github-account** [comment]
@markshevchenko 
Мне видится следующая реализация:
1) в предметной области в класс Order добавить метод для удаления определенного кол-ва книг и метод для удаления всех книг из заказа, сделать для них соответствующие контроллеры.
2) на уровне представления добавить кнопки: 
    - "Удалить все". Удаляет все экземпяляры данной книги.
    - Кнопки "+" и "-". Регулируют кол-во экземпляров данной книги.

Открытый вопрос: как отправлять с клиента данные об изменнии кол-ва экземпляров книги в корзине? Это должна быть или форма, или AJAX-запросы при нажатии на кнопки "+" "-" и "Удалить все"?

**Comment by markshevchenko** [comment]
> в предметной области в класс Order добавить метод для удаления определенного кол-ва книг и метод для удаления всех книг из заказа, сделать для них соответствующие контроллеры.

Да. Только новых контроллеров не нужно, все методы логично разместить в уже существующем контроллере `OrderController`.

> на уровне представления добавить кнопки:
> 
> "Удалить все". Удаляет все экземпяляры данной книги.
> Кнопки "+" и "-". Регулируют кол-во экземпляров данной книги.

Тоже да.

> Открытый вопрос: как отправлять с клиента данные об изменнии кол-ва экземпляров книги в корзине? Это должна быть или форма, или AJAX-запросы при нажатии на кнопки "+" "-" и "Удалить все"?

В соответствии с принципом [Responsive web design](https://www.w3schools.com/html/html_responsive.asp), который по-русски называют [Адаптивным веб-дизайном](https://ru.wikipedia.org/wiki/Адаптивный_веб-дизайн), сайт надо разрабатывать так, чтобы он был работоспособен на разных устройствах и в разных условиях. В частности, надо обеспечить его работу с отключенным JavaScript.

Для нас это означает, что сначала мы делаем всё на чистых формах (сейчас именно так и делается), а, если останется время, можно добавить AJAX. Это не очень трудно в ASP.NET Core.

**Comment by lukas-github-account** [comment]
@markshevchenko 
Если происходит попытка удалить книгу из пустой корзины, нужно ли в этом случае бросать исключение? Или достаточно проверки, что корзина пуста, и если да, просто сделать return из метода.

**Comment by markshevchenko** [comment]
> Если происходит попытка удалить книгу из пустой корзины, нужно ли в этом случае бросать исключение? Или достаточно проверки, что корзина пуста, и если да, просто сделать return из метода.

В нашем случае мы сами решаем. У нас учебных проект, поэтому я предлагаю сделать с исключением. Когда будем разбираться с обработкой ошибок, будет материал для работы. Такую ошибку неплохо логгировать.

**Comment by lukas-github-account** [comment]
@markshevchenko 
Некоторые проверки и выброс исключения дублируются в методах для удаления определенного кол-ва книг и для удаления всех книг из заказа. Например, проверка, что корзина не пустая. Уместно ли такое дублирование кода в данной ситуации?

**Comment by markshevchenko** [comment]
@executab1e, я бы вообще отказался от метода удаления определённого количества книг. Лучше идти от историй при проектировании возможностей.
Обычно в магазинах у покупателя есть возможность удалять книги из-за заказа по одной. Это значит, что в конечном итоге у нас будет метод наподобие `Order.RemoveItem(int bookId)`, а метод удаления сразу нескольких книг не следуюет ни из какой бизнес-истории, а, следовательно, не нужен.

**Comment by lukas-github-account** [comment]
Да, согласен.
